### ISSUE TITLE
Corrupted frame on the edge 

### ISSUE THREAD

**Issue by mhtrinh** [mikeboers/PyAV#423]
I am using version 0.5.3 (pip install)
I have a H264 video, size 1294x964 

When decoding I get frame of `<av.VideoFormat yuv420p, 1294x964>`
The problem is when I convert it to image (`frame.to_image()` or `frame.to_ndarray(format='bgr24')`) the resulting image have the last 6 pixel columns fill with random values.
Is it because my width is not a multiple of 8 ? 
I don't have this issue when using ffmpeg in C, where I use swscale to convert YUV to BGR. 
Is there anyway to go around this problem ?

**Comment by mhtrinh** [comment]
Please find the link https://www.dropbox.com/sh/zl3j7gaqpk1xdtx/AAAhQM5vq3sbpi7eRT1gA4YKa?dl=0 to :
- mp4 of 500x500 
- ffmpegExtract : frames extracted using ffmpeg command line
- pyavExtract : frames extracted using below code with PyAV : the last 4 column are random noise (not on every frame as it's random ...)

PyAV frame extract code :
```
import av

path="/home/mhtrinh/temp/numberedFrame/builds/Debug/numberedTo100.mp4"
container=av.open(path)
stream=container.streams[0]

def  nextFrame(container,stream):
    for packet in container.demux(stream):
        for frame in packet.decode():
            yield frame
            
for i in range(10):
    frame=next(nextFrame(container,stream))
    fn="img.{:03d}.png".format(i)
    frame.to_image().save(fn)
    print(fn)
```

**Comment by jlaine** [comment]
For to_ndarray this is almost certainly fixed by #415.

**Comment by jlaine** [comment]
OK, I'm going to set aside `to_image` for now as it definitely doesn't deal with alignment properly.

I found a thinko in #415 which I have fixed in 93e56148b094b10cf3d97b7168958cfe48ef1d06, so that `.to_ndarray()` actually works with frames which have some padding.

With `develop` the following works perfectly with the video provided by @mhtrinh :

```
import av
import cv2

container = av.open('numbered_500x500_problematicVideo.mp4')
stream = container.streams[0]

for i, frame in enumerate(container.decode(stream)):
    fn="{:03d}.pyav.png".format(i + 1)
    arr = frame.to_ndarray()
    arr = cv2.cvtColor(arr, cv2.COLOR_YUV2BGR_I420)
    cv2.imwrite(fn, arr)
```

This on the other hand still gives me corruption (visible on frames 17-18-19 for @mhtrinh's video, but I see the same on other videos):

```
import av
import cv2

container = av.open('numbered_500x500_problematicVideo.mp4')
stream = container.streams[0]

for i, frame in enumerate(container.decode(stream)):
    fn="{:03d}.pyav.png".format(i + 1)
    arr = frame.to_ndarray(format='bgr24')
    cv2.imwrite(fn, arr)
```

If you look at the difference between these two pieces of code it's that in the first we extract the YUV420P data directly and let OpenCV do the conversion to BGR, and in the second we make use of PyAV's `reformat`. This suggests to me that `reformat` is not working as expected.

**Comment by jlaine** [comment]
I *think* it's got something to do with the alignment of the output frame we pass to `sws_scale`.

In @mhtrinh's case we have yuv420p input provided by the codec with:

- Y : 500x500 pixels, with a linesize of 512
- U & V : 255x255pixels, with a linesize of 256

Our output is a single BGR packed plane of 500x500 pixels with a linesize of 1500. I've tried playing with the "align" value of the output frame we allocate, and the first value for which I get perfect output sems to be 64. I don't understand this, I've read messages which mention being 32-byte aligned, but not 64.

**Comment by jlaine** [comment]
The root cause was fixed in ad15d97ddf194af0a009f78607b92bd22fc763f5
